clients/application-amqp/src/main/java/org/eclipse/hono/application/client/amqp/ProtonBasedRequestResponseCommandClient.java

```diff
@@ -119,28 +124,57 @@ protected BufferResult getResult(final int status, final String contentType, fin
+                            connection.getConfig());
                     return client.createAndSendRequest(command, messageTargetAddress, properties, data, contentType,
-                            this::getRequestResponseResult, currentSpan);
+                            this::mapCommandResponse, currentSpan);
                 })
                 .recover(error -> {
                     Tags.HTTP_STATUS.set(currentSpan, ServiceInvocationException.extractStatusCode(error));
                     TracingHelper.logError(currentSpan, error);
                     return Future.failedFuture(error);
                 })
-                .map(bufferResult -> {
-                    setTagsForResult(currentSpan, bufferResult);
-                    if (bufferResult != null && bufferResult.isError()) {
-                        throw StatusCodeMapper.from(bufferResult);
+                .map(result -> {
+                    if (result == null) {
+                        throw new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST);
+                    } else {
+                        setTagsForResult(currentSpan, result);
+                        if (result.isError()) {
+                            throw StatusCodeMapper.from(result);
+                        }
+                        return result.getPayload();
                     }
-                    return bufferResult;
                 })
                 .onComplete(r -> currentSpan.finish());
     }
 
-    private Future<RequestResponseClient<BufferResult>> getOrCreateClient(final String tenantId, final String replyId) {
+    private RequestResponseResult<DownstreamMessage<AmqpMessageContext>> mapCommandResponse(final Message message,
+            final ProtonDelivery delivery) {
+        final DownstreamMessage<AmqpMessageContext> downStreamMessage = ProtonBasedDownstreamMessage.from(message, delivery);
+
+        return Optional
+                .ofNullable(MessageHelper.getStatus(message))
+                .map(status -> new RequestResponseResult<>(status, downStreamMessage,
+                        CacheDirective.from(MessageHelper.getCacheDirective(message)), null))
+                .orElseGet(() -> {
+                    LOGGER.trace(
+                            "response message has no status code application property [reply-to: {}, correlation ID: {}]",
+                            message.getReplyTo(), message.getCorrelationId());
+                    return null;
+                });
+    }
+
+    @Override
+    protected RequestResponseResult<DownstreamMessage<AmqpMessageContext>> getResult(final int status,
+            final String contentType, final Buffer payload, final CacheDirective cacheDirective,
+            final ApplicationProperties applicationProperties) {
+        // This method is not used but need to be overridden as it is defined as abstract in the parent class
+        return null;
```

Comment:
> Another solution could even be to let this method actually be used and remove the `mapCommandResponse` method. This would mean removing the new extra `RequestResponseClient.createAndSendRequest` method and adapting the existing `RequestResponseClient.createAndSendRequest` methods to use the `BiFunction<Message, ProtonDelivery, R>` `responseMapper` param with the added ProtonDelivery. This would also mean changing the `getResult` method to also contain the delivery, potentially via an extra `Pair<Message, ProtonDelivery>` parameter (which would then be ignored in most implementations). WDYT?

Initially I started with such an approach. There are lot of existing classes getting affected with that and the PR was quite large. Hence I chosen this approach. Adding an extra `Pair<Message, ProtonDelivery>` message seems to me not optimal as  other parameters such as `status`, `contentType`, `payload` and `applicationProperties`  are already part of the `message`. For this PR I would prefer to go with throwing an `UnsupportedOperationException`. We can still refactor these classes after this PR.

Reply:
I have updated with `UnsupportedOperationException` and pushed a new commit for it.

Reply:
> For this PR I would prefer to go with throwing an `UnsupportedOperationException`. We can still refactor these classes after this PR.

Ok